server/src/main/java/org/opensearch/index/IndexModule.java

```diff
@@ -672,4 +682,23 @@ public static Map<String, IndexStorePlugin.DirectoryFactory> createBuiltInDirect
         }
         return factories;
     }
+
+    /**
+     * Adds a supplier of extension's list. All listeners added here are maintained for the entire index lifecycle on
+     * this node. Once an index is closed or deleted these listeners go out of scope.
+     * <p>
+     * Note: an index might be created on a node multiple times. For instance if the last shard from an index is
+     * relocated to another node the internal representation will be destroyed which includes the registered listeners.
+     * Once the node holds at least one shard of an index all modules are reloaded and listeners are registered again.
+     * Listeners can't be unregistered they will stay alive for the entire time the index is allocated on a node.
+     * </p>
+     */
+    public void addFSTypeFileExtensions(final Map<Type, Set<String>> extensionsByType) {
```

Comment:
The propagation of the FS extensions is not clear to me, who will call this method and when? I see it only being used in tests ...

Reply:
It's the plugin who needs to register additional extensions, it has IndexModule instance which is available via [Plugin class](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/Plugin.java#L184) and this will be part of initialization. It's similar to what plugin do today to register event listeners, e.g. k-NN has [custom listener](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/KNNSettings.java#L476) for SettingsUpdate event. 

So method call in plugin will be something like:

```
public void onIndexModule(IndexModule indexModule) {
        indexModule. addFSTypeFileExtensions(Map.of(IndexModule.Type.HYBRIDFS, Set.of("ext1", "ext2")));
 }
```

Reply:
Excellent, so what we could try here is to leverage `IndexSettings`, something along these lines (the functionality does not exist yet):

```
public void onIndexModule(IndexModule indexModule) {
        indexModule.updateIndexSetting (Module.INDEX_STORE_HYBRID_MMAP_EXTENSIONS,  Set.of("ext1", "ext2")),
          ... merge function);
 }

```

In this case you could get rid of all FS specific methods, `additionlaExtensions` arguments and purely rely on existing `IndexSettings` mechanism.

Reply:
I see you idea, basically update the setting for hybridFs extensions and all the values will flow to FsFactory class without any additional changes. One concern that I have is that we can freely update any setting, I may need to test it ([with this public method](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java#L823)) but I'm not sure it's possible with non-dynamic settings. Currently that [hybridFsExtensions](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/IndexModule.java#L159) is non-dynamic setting.
Also such calls will trigger any registered onSettingUpdate listeners, and maybe will expose setting for API's update (this is in case we need to make it dynamic). Just want to make sure this is acceptable.

Reply:
> One concern that I have is that we can freely update any setting, I may need to test it ([with this public method](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java#L823)) but I'm not sure it's possible with non-dynamic settings.

Correct, the [hybridFsExtensions](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/IndexModule.java#L159) is not dynamic and probably should not be (since once FS is initialized, it could not be changed on the fly).

> Also such calls will trigger any registered onSettingUpdate listeners

That could happen but I don't see update listener for `hybridFsExtensions` (since this is not dynamic).  I have just this general idea which we could work through but the reasoning behind is - FS code should not leak anywhere (like it does now), `IndexSettings` is the only source for index configuration.

Reply:
@reta  I see a problem with a Settings approach. If setting is not dynamic then I can't register customer update handler (similar to other [existing settings](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/IndexSettings.java#L831)) to perform "append"-like update and combine existing core and custom extensions into one list. With default updater new value overwrite existing ones, which normally should be enough but not in this case. The problem is in[ this validation](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/settings/Setting.java#L673) that checks if setting is dynamic.

Reply:
@martin-gaievski I think I found exactly what we need, each `Plugin` has the way to contribute additional `IndexSettings`, it may need some minor tweaks (I can help with that), but I believe it should address all the concerns raised:

```
    /**
     * An {@link IndexSettingProvider} allows hooking in to parts of an index
     * lifecycle to provide explicit default settings for newly created indices. Rather than changing
     * the default values for an index-level setting, these act as though the setting has been set
     * explicitly, but still allow the setting to be overridden by a template or creation request body.
     */
    public Collection<IndexSettingProvider> getAdditionalIndexSettingProviders() {
        return Collections.emptyList();
    }
```

Reply:
Inspired by your finding, I tested even simpler method 

```
    /**
     * Additional node settings loaded by the plugin. Note that settings that are explicit in the nodes settings can't be
     * overwritten with the additional settings. These settings added if they don't exist.
     */
    public Settings additionalSettings() {
        return Settings.Builder.EMPTY_SETTINGS;
    }
```

It works for our setting, just only one thing is that it overrides value in core, so I had to do read-update. 

```
public Settings additionalSettings() {
        final List<String> knnSpecificSettings = List.of("vem", "vec");
        final List<String> combinedSettings = Stream.concat(
            IndexModule.INDEX_STORE_HYBRID_MMAP_EXTENSIONS.getDefault(Settings.EMPTY).stream(),
            knnSpecificSettings.stream()
        ).collect(Collectors.toList());
        return Settings.builder().putList(IndexModule.INDEX_STORE_HYBRID_MMAP_EXTENSIONS.getKey(), combinedSettings).build();
    }
```

I think that's ok as core allowing it in the first place. With this all changes are on plugin side, I'll be closing this PR. Thank you @reta